### ISSUE TITLE
publish releases to maven

### ISSUE THREAD

**Issue by rursprung** [aiven/prometheus-exporter-plugin-for-opensearch#64]
it'd be great if you could publish your releases to maven (we use it to manage our dependencies, incl. opensearch & opensearch plugins).

you can use the `opensearch.pluginzip` gradle plugin for this: see the [opensearch plugins documentation on this](https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#opensearchpluginzip) and https://github.com/opensearch-project/opensearch-build/issues/1916.

**Comment by lukas-vlcek** [comment]
Hi @rursprung 

If I read it correctly you would like to have the `apply plugin: 'opensearch.pluginzip'` be added to `build.gradle` so that you can call the new custom publish task `publishPluginZipPublicationToZipStagingRepositoryand` to have the plugin ZIP be pushed into mvn repo of your choice?

Do I get correctly?

**Comment by rursprung** [comment]
not quite: the idea would be that the publish would happen as part of your normal release, so that it's on maven central (or any other public maven repo). then it can be pulled in by everyone who needs it (or manually downloaded; same as through the GH releases page - that way it can act as a mirror as well).
but yes, this happens - partially - through that plugin. i haven't used that plugin, so i'm not 100% sure how much of the publishing it automates away, but it's being added to all plugins under the opensearch organisation (some have already done it, others still have an open ticket for it).

**Comment by lukas-vlcek** [comment]
I think this is also related: https://github.com/opensearch-project/OpenSearch/issues/3692

**Comment by lukas-vlcek** [comment]
@rursprung based on the discussion in https://github.com/opensearch-project/OpenSearch/issues/3692 I will "friendly" close this ticket as won't fix. Once OpenSearch enables custom groupId values in POM we can revisit it.

**Comment by lukas-vlcek** [comment]
Status update:
- A blog post about the ZIP pom publication plugins has been published: https://opensearch.org/blog/technical-post/2022/08/opensearch-plugin-zips-now-in-maven-repo/
- A PR to allow maven pom `groupId` customization is being reviewed: https://github.com/opensearch-project/OpenSearch/pull/4156

**Comment by lukas-vlcek** [comment]
Update:

- The Zip publication POM `groupId` customization will be available starting with OpenSearch release v2.4: https://github.com/opensearch-project/opensearch-build/issues/2521

Once the [OpenSearch release 2.4](https://github.com/opensearch-project/opensearch-build/issues/2649) is out we can re-open this ticket.

**Comment by rursprung** [comment]
@lukas-vlcek 2.4 is out, i guess this can be re-opened? (not something that i can do) 🙂 
thanks a lot for all your heavy work on the publishing mechanism!

**Comment by lukas-vlcek** [comment]
Sure, we can now continue.

I think one of the changes I would like to do first is changing the "package" name of internal classes to something that will match the groupId in the maven repo. Perhaps to something like `io.aiven.plugin` or similar. Definitely not `org.opensearch` because that is already taken by the core OpenSearch plugins.

_(Although it is possible to override the `groupId` in gradle publication task I would rather use package name that is really used by the plugin in the code, hence the change ^^)_

**Comment by rursprung** [comment]
@lukas-vlcek: do you think it'd now be possible to pick this up? the packages are still named `org.opensearch.[..]`, not sure if you've already picked a new name?
thanks!

**Comment by dancristiancecoi** [comment]
Hello, any plans for this to be worked on in the near future?
Btw, thanks a lot for releasing and maintaining this project!

**Comment by lukas-vlcek** [comment]
Right, I will start working on that too.

It will need to become part of the release process, perhaps this article: [Publishing Java packages with Maven](https://docs.github.com/en/actions/publishing-packages/publishing-java-packages-with-maven) is a good starting point.

**Comment by rursprung** [comment]
now that #489 has been implemented i think that the only thing missing would be to publish the artefacts to maven central as part of the release process.
but i don't know what the release process for this plugin is (i didn't see a GitHub CI job for it => it might be done manually or runs on other infrastructure?). since this is being done for other plugins in the same organisation (e.g. `opensearch-security`) the same approach as for them can/should probably be taken to publish releases. i'm not sure who can provide the necessary insights for this
